ooni/oonicli.py

```diff
@@ -315,3 +329,231 @@ def start():
+            sys.exit(2)
+        except errors.NetTestNotFound as path:
+            log.err('Requested NetTest file not found (%s)' % path)
+            sys.exit(3)
+        except errors.OONIUsageError as e:
+            log.err(e)
+            print e.net_test_loader.usageOptions().getUsage()
+            sys.exit(4)
+        except Exception as e:
+            if config.advanced.debug:
+                log.exception(e)
+            log.err(e)
+            sys.exit(5)
+
+        d = director.start(start_tor=True,
+                           check_incoherences=check_incoherences)
+
+        def setup_nettest(_):
+            try:
+                return deck.setup()
+            except errors.UnableToLoadDeckInput as error:
+                return defer.failure.Failure(error)
+
+
+
+        # Wait until director has started up (including bootstrapping Tor)
+        # before adding tests
+        def post_director_start(_):
+            for net_test_loader in deck.netTestLoaders:
+                # Decks can specify different collectors
+                # for each net test, so that each NetTest
+                # may be paired with a test_helper and its collector
+                # However, a user can override this behavior by
+                # specifying a collector from the command-line (-c).
+                # If a collector is not specified in the deck, or the
+                # deck is a singleton, the default collector set in
+                # ooniprobe.conf will be used
+
+                collector = None
+                if not global_options['no-collector']:
+                    if global_options['collector']:
+                        collector = global_options['collector']
+                    elif 'collector' in config.reports \
+                            and config.reports['collector']:
+                        collector = config.reports['collector']
+                    elif net_test_loader.collector:
+                        collector = net_test_loader.collector
+
+                if collector and collector.startswith('httpo:') \
+                        and (not (config.tor_state or config.tor.socks_port)):
+                    raise errors.TorNotRunning
+
+                test_details = net_test_loader.testDetails
+                test_details['annotations'] = global_options['annotations']
+
+                director.startNetTest(net_test_loader,
+                                      global_options['reportfile'],
+                                      collector)
+            return director.allTestsDone
+
+        def start():
+            d.addCallback(setup_nettest)
+            d.addCallback(post_director_start)
+            d.addErrback(director_startup_handled_failures)
+            d.addErrback(director_startup_other_failures)
+            return d
+
+        return start()
+
+
+    finished = defer.Deferred()
+
+    @defer.inlineCallbacks
+    def readmsg(_,channel, queue_object, consumer_tag):
```

Comment:
I would perhaps refactor this function into a separate object so that you don't have to use a global counter and lifetime, but they can simply be part of the object attributes.